### PR TITLE
Change move constructor and assignment operator for Memory<T>.

### PR DESCRIPTION
**Pull request by YohannDudouit**
The move semantic expect behavior is to take ownership, this PR fixes this behavior for the `Memory<T>` class.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#2355](https://github.com/mfem/mfem/pull/2355) | @YohannDudouit | @tzanio | @acfisher + @v-dobrev + @pazner + @camierjs | 6/26/21 | 4/9/22 | 4/15/22 | |

### CODE REVIEW

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

File: general/mem_manager.hpp [hunk 1]
```diff
@@ -175,14 +175,28 @@ class Memory
    /// Copy constructor: default.
    Memory(const Memory &orig) = default;
 
-   /// Move constructor: default.
-   Memory(Memory &&orig) = default;
+   /// Move constructor, takes ownership when possible and invalidates @a orig.
```
Comment (pazner) [inline general/mem_manager.hpp:178]: What does it mean "when possible"?

File: general/mem_manager.hpp [hunk 2]
```diff
@@ -175,14 +175,28 @@ class Memory
...
+   Memory(Memory &&orig)
+      : h_ptr(orig.h_ptr), capacity(orig.capacity), h_mt(orig.h_mt),
+        flags(orig.flags)
+   {
+      orig.Reset();
+   }
```
Comment (pazner) [inline general/mem_manager.hpp:184]:
```suggestion
   Memory(Memory &&orig)
   {
      *this = orig;
      orig.Reset();
   }
```

File: general/mem_manager.hpp [hunk 3]
```diff
@@ -175,14 +175,28 @@ class Memory
...
+       @a orig. */
+   Memory &operator=(Memory &&orig)
+   {
+      h_ptr = orig.h_ptr;
+      capacity = orig.capacity;
+      h_mt = orig.h_mt;
+      flags = orig.flags;
+      orig.Reset();
+      return *this;
+   }
```
Comment (pazner) [inline general/mem_manager.hpp:199]:
```suggestion
   Memory &operator=(Memory &&orig)
   {
      *this = orig;
      orig.Reset();
      return *this;
   }
```

**Review by pazner — COMMENTED** [review]
The change here is that the move constructor invalidates (clears ownership flags) of the moved `Memory`.

It is technically not required by C++ move semantic — the moved object can be left in any state, as long as it is destructible, and the destructor `~Memory` is trivial.

That being said, it may be less confusing since it will avoid having two `Memory` objects both with ownership flags set.

File: general/mem_manager.hpp (hunk 1 above)
Comment (YohannDudouit) [reply general/mem_manager.hpp:178]: "when `@a orig` has ownership", would you prefer that in the documentation?

Comment (pazner) [reply general/mem_manager.hpp:178]: Maybe something like
```c++
/// Move constructor. Sets the pointers and associated ownership of validity
/// flags of @a *this to those of @a other. Invalidates @a other.
```

File: general/mem_manager.hpp (hunk 2 above)
Comment (YohannDudouit) [reply general/mem_manager.hpp:184]: What I don't like with this implementation is that it relies on the copy assignment operator to create "multiple" owners which I don't like and would like to change in the future since it is bug prone (double `free`, usage of data after `free`, etc...).
I'm also wondering if we should call `Delete`, since it's potentially a memory leak not to, and I can't think of a good reason not to call `Delete`.

File: general/mem_manager.hpp (hunk 3 above)
Comment (YohannDudouit) [reply general/mem_manager.hpp:199]: ditto.

File: general/mem_manager.hpp (hunk 2 above)
Comment (pazner) [reply general/mem_manager.hpp:185]: Both versions are functionally the same (there are two owners immediately before `orig.Reset()` in both versions of the code). The advantage of what I suggested is that it should work without modification even if you add more data members.

It is not the role of the move constructor to destroy the object. Its destructor will be called eventually (usually very soon).

Comment (YohannDudouit) [reply general/mem_manager.hpp:185]: What I meant is that the version you're proposing will become wrong if we change the behavior of the copy. But I'm okay using the copy to write the move if it's important to you.

I don't think you can discard calling `Delete()` that easily, you have to remember that the destructor does not call `Delete()`, so using move semantics without calling first `Delete()` would result in a memory leak, and I do not see any benefits in letting that ever happen.

**Review by pazner — APPROVED** [review]

**Comment by v-dobrev** [comment]
As with the other PR, #2358, this PR goes against my original design. I prefer not to make this change, so I will not review this PR.

**Comment by YohannDudouit** [comment]
As @pazner was mentioning during the team meeting there is probably different valid choices for the move semantics behavior. This version leaves the moved objects in a "Reset" state, meaning that the origin is unusable after being moved (I think this is the expected standard behavior). An alternate behavior could just be to make a copy of the origin __and__ to take ownership of the content of the origin object when possible (the origin is still usable but does not own the memory any more, and can therefore be deleted __safely__).

I don't think the move semantics is used anywhere so far, but in a way I agree with @tzanio that the behavior of the move semantics should be coherent with the behavior of the copy constructors. In this regard, I think that the only meaningful behavior for the copy constructor is to copy __without__ taking ownership (otherwise dangling pointer bugs become very likely, or double `free` issues, since copying creates multiple owners). Ownership issues is one of the most frequent things that our users complain about, and for this reason I don't think we should take this very computer _science_ question too lightly. I want to remind that the __ownership__ aspect of the `Memory<T>` class goes beyond what pointers do, pointers always own their memory (they can always delete it, and create dangling pointers, double `free`, etc..). In many regards pointers are a bad abstraction of memory because they enable all sort of undesired features (bugs), modern C++ guidelines recommend to __never__ use raw pointers (There is always a better solution provided by the standard library). There is __absolutely no benefits__ in having a design that allows dangling pointers. A last remark, this design doesn't fully get rid of dangling pointers issues, since the "origin" is not aware of its aliases, a user can delete the "origin" and leave all aliases dangling. Making aliases __safe__ would require a more extensive redesign.

To summarize, if we want the move semantics to mean "copy and take ownership", and the copy constructor to mean "copy and don't take ownership", it may make more sense to have the move semantics leave the "origin" in a valid state (contrary to what is done in this PR).

**Comment by tzanio** [comment]
@YohannDudouit, am I correct that this is waiting on some documentation updates that we have discussed previously?

File: general/mem_manager.hpp [hunk 4]
```diff
@@ -106,11 +106,22 @@ MemoryClass operator*(MemoryClass mc1, MemoryClass mc2);
     In many respects this class behaves like a pointer:
     * When destroyed, a Memory object does NOT automatically delete any
       allocated memory.
-    * Only the method Delete() will deallocate a Memory object.
-    * Other methods that modify the object (e.g. New(), Wrap(), etc) will simply
-      overwrite the old contents.
-    * One difference with a pointer is that a const Memory object does not allow
-      modification of the content (unlike e.g. a const pointer).
+    * Only the method `Delete()` will deallocate a Memory object.
```
Comment (acfisher) [inline general/mem_manager.hpp:109]: Thanks for adding the extensive documentation on this.

**Review by acfisher — APPROVED** [review]

**Comment by tzanio** [comment]
@v-dobrev can you confirm that your documentation concerns are addressed now?

**Comment by tzanio** [comment]
Merged in `next` for testing...

File: general/mem_manager.hpp [hunk 5]
```diff
@@ -112,11 +112,22 @@ MemoryClass operator*(MemoryClass mc1, MemoryClass mc2);
...
```
Comment (v-dobrev) [inline general/mem_manager.hpp:115]: Typically, when we refer to methods/classes (which auto-generates links), we do not put them in back-quotes, see e.g. here: https://github.com/mfem/mfem/blob/3c1859e1d7d9d7bb35ed3a5261f15dcc24343d9a/linalg/operator.hpp#L65-L66

This is consistent with how Doxygen displays the methods/classes themselves: see e.g. the "Public Member Functions" section at https://docs.mfem.org/4.4/classmfem_1_1Operator.html

Comment (v-dobrev) [reply general/mem_manager.hpp:115]: That said, this section is actually not formatted as a bullet list by doxygen -- to get that we need to re-format this a bit, namely replace `*` with `-` since the first `*` in this type of comments is ignored.

File: general/mem_manager.hpp [hunk 6]
```diff
@@ -112,11 +112,22 @@ MemoryClass operator*(MemoryClass mc1, MemoryClass mc2);
...
+    In other aspects this class differs from a pointer:
+    * Pointer arithmetic is not supported, `MakeAlias()` should be used instead.
+    * Const Memory object does not allow modification of the content
+      (unlike e.g. a const pointer).
+    * Move constructor and assignement will transfer ownership flags, and
+      `Reset()` the moved Memory object.
+    * Copy constructor and assignement copy flags. This may result in two Memory
+      objects owning the data which is an invalid state. This invalid state MUST
+      be resolved by users manually using `SetHostPtrOwner()`,
+      `SetDevicePtrOwner()`, or `ClearOwnerFlags()`.
```
Comment (v-dobrev) [inline general/mem_manager.hpp:127]: One can also just NOT delete the copy (or the original).

File: general/mem_manager.hpp [hunk 7]
```diff
@@ -112,11 +112,22 @@ MemoryClass operator*(MemoryClass mc1, MemoryClass mc2);
...
+    * When creating alias memory objects, the consistency of memory flags have
+      to be manually taken care of using either `Sync()` or `SyncAlias`. Failure
+      to do so will result in silent misuse of unsynchronized data.
```
Comment (v-dobrev) [inline general/mem_manager.hpp:130]: To be more precise: "When moving or copying (between host and device) alias Memory objects and/or their base Memory objects, ...."

File: general/mem_manager.hpp [hunk 8]
```diff
@@ -182,14 +193,25 @@ class Memory
    /// Copy constructor: default.
    Memory(const Memory &orig) = default;
 
-   /// Move constructor: default.
-   Memory(Memory &&orig) = default;
+   /** Move constructor. Sets the pointers and associated ownership of validity
+       flags of @a *this to those of @a other. Invalidates @a other. */
```
Comment (v-dobrev) [inline general/mem_manager.hpp:197]: Instead of "Invalidates", I'd just say "Resets" or "Clears the contents of".

File: general/mem_manager.hpp [hunk 9]
```diff
@@ -182,14 +193,25 @@ class Memory
...
+      orig.Reset();
+   }
 
    /// Copy-assignment operator: default.
    Memory &operator=(const Memory &orig) = default;
 
-   /// Move-assignment operator: default.
-   Memory &operator=(Memory &&orig) = default;
+   /** Move assignment operator. Sets the pointers and associated ownership of
+       validity flags of @a *this to those of @a other. Invalidates @a other. */
```
Comment (v-dobrev) [inline general/mem_manager.hpp:208]: Similar here: "Invalidates" --> "Resets" or "Clears the contents of".

File: general/mem_manager.hpp (hunk 6 above)
Comment (v-dobrev) [reply general/mem_manager.hpp:127]: I meant that it might be good to add the above sentence to this bullet.

File: general/mem_manager.hpp (hunk 7 above)
Comment (v-dobrev) [reply general/mem_manager.hpp:130]: "SyncAlias" --> "SyncAlias()"

File: general/mem_manager.hpp (hunk 5 above)
Comment (YohannDudouit) [reply general/mem_manager.hpp]: done in [418054b](https://github.com/mfem/mfem/pull/2355/commits/418054b770a4cac104a33b46f18dce432a3f5482)

File: general/mem_manager.hpp (hunk 7 above)
Comment (YohannDudouit) [reply general/mem_manager.hpp]: done in [418054b](https://github.com/mfem/mfem/pull/2355/commits/418054b770a4cac104a33b46f18dce432a3f5482)

File: general/mem_manager.hpp (hunk 6 above)
Comment (YohannDudouit) [reply general/mem_manager.hpp]: done in [418054b](https://github.com/mfem/mfem/pull/2355/commits/418054b770a4cac104a33b46f18dce432a3f5482)

File: general/mem_manager.hpp (hunk 8 above)
Comment (YohannDudouit) [reply general/mem_manager.hpp]: done in [418054b](https://github.com/mfem/mfem/pull/2355/commits/418054b770a4cac104a33b46f18dce432a3f5482)

File: general/mem_manager.hpp (hunk 9 above)
Comment (YohannDudouit) [reply general/mem_manager.hpp]: done in [418054b](https://github.com/mfem/mfem/pull/2355/commits/418054b770a4cac104a33b46f18dce432a3f5482)

**Comment by tzanio** [comment]
Re-merged in `next` for testing...

**Review by v-dobrev — APPROVED** [review]

File: general/mem_manager.hpp [hunk 10]
```diff
@@ -110,13 +110,28 @@ MemoryClass operator*(MemoryClass mc1, MemoryClass mc2);
...
+      (unlike e.g. a const pointer).
+    - Move constructor and assignement will transfer ownership flags, and
+      Reset() the moved Memory object.
+    - Copy constructor and assignement copy flags. This may result in two Memory
+      objects owning the data which is an invalid state. This invalid state MUST
+      be resolved by users manually using SetHostPtrOwner(),
+      SetDevicePtrOwner(), or ClearOwnerFlags(). One can also just NOT use
+      Delete() on one of the two Memory objects, however this is discouraged as
+      it is counterproductive to have developpers be responsible for tracking
+      ownership when flags are dedicated to this task.
```
Comment (tzanio) [inline general/mem_manager.hpp:130]: Can you clarify this sentence? I can't understand it as written

"One can also just NOT use Delete() on one of the two Memory objects, however this is discouraged as it is counterproductive to have developpers be responsible for tracking ownership when flags are dedicated to this task."

Also, I will replace "as it is counterproductive to have developers be responsible for tracking ownership when flags are dedicated to this task" with "since ownership is tracked by internal flags".

Comment (YohannDudouit) [reply general/mem_manager.hpp]: done in [b3a1202](https://github.com/mfem/mfem/pull/2355/commits/b3a120277a1fb9e083cc06f651021294fa8d76b6)

**Review by tzanio — APPROVED** [review]
